src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -245,12 +246,116 @@ private BoundExpression ApplyConvertedTypes(BoundExpression expr, TupleBinaryOpe
         }
 
+        /// <summary>
+        /// If an element in a tuple literal has an explicit name which doesn't match the name on the other side, we'll warn.
+        /// The user can either remove the name, or fix it.
+        /// 
+        /// This method handles two expressions, each of which is either a tuple literal or an expression with tuple type.
+        /// In a tuple literal, each element can have an explicit name, an inferred name or no name.
+        /// In an expression of tuple type, each element can have a name or not.
+        /// </summary>
+        private void ReportNamesMismatchesIfAny(BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            bool leftIsTupleLiteral = left.Kind == BoundKind.TupleLiteral;
+            bool rightIsTupleLiteral = right.Kind == BoundKind.TupleLiteral;
+
+            if (!leftIsTupleLiteral && !rightIsTupleLiteral)
+            {
+                return;
+            }
+
+            var (leftNames, leftInferred, leftLocations) = getDetails(left);
+            bool leftNoInferredNames = leftInferred.IsDefault;
+            bool leftNoNames = leftNames.IsDefault;
+
+            var (rightNames, rightInferred, rightLocations) = getDetails(right);
+            bool rightNoInferredNames = rightInferred.IsDefault;
+            bool rightNoNames = rightNames.IsDefault;
+
+            if (leftNoNames && rightNoNames)
+            {
+                return;
+            }
+
+            Debug.Assert(leftNoNames || rightNoNames || leftNames.Length == rightNames.Length);
+
+            // The only case where the right side is not preferred is when we have a tuple literal on the left, but not the right
+            bool isRightBestSide = !(leftIsTupleLiteral && !rightIsTupleLiteral);
+
+            int length = leftNoNames ? rightNames.Length : leftNames.Length;
+            for (int i = 0; i < length; i++)
+            {
+                string leftName = leftNoNames ? null : leftNames[i];
+                bool leftWasInferred = leftNoInferredNames ? false : leftInferred[i];
+
+                string rightName = rightNoNames ? null : rightNames[i];
+                bool rightWasInferred = rightNoInferredNames ? false : rightInferred[i];
+
+                bool different = string.CompareOrdinal(rightName, leftName) != 0;
+
+                bool leftComplaint = leftIsTupleLiteral && leftName != null && !leftWasInferred && (rightNoNames || different);
+                bool rightComplaint = rightIsTupleLiteral && rightName != null && !rightWasInferred && (leftNoNames || different);
+
+                if (!leftComplaint && !rightComplaint)
+                {
+                    // No complaints, let's move on
+                    continue;
+                }
+
+                BoundExpression location = null;
+                string complaintName = null;
+                if (leftComplaint && rightComplaint)
```

Comment:
Can we simplify down to a single case:
```
bool useRight = rightComplaint && (!leftComplaint || isRightBestSide);
var location = useRight ? rightLocations[i] : leftLocations[i];
var compaintName = useRight ? rightName : leftName;
```

Reply:
Is it simpler? At least for me, it's easier to tell if the code as-is is correct, while the compact code is harder to tell (I'm not sure).

---
In reply to: [173891376](https://github.com/dotnet/roslyn/pull/25396#discussion_r173891376) [](ancestors = 173891376)

Reply:
I still can't follow that one-liner. Changed it in a way that I think is a bit easier to follow. Let me know what you think